src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -0,0 +1,357 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNet.OData.Builder;
+using Microsoft.AspNet.OData.Common;
+using Org.OData.Core.V1;
+
+namespace Microsoft.AspNet.OData
+{
+    /// <summary>
+    /// Represents an <see cref="IDeltaSet"/> that is a collection of <see cref="IDeltaSetItem"/>s.
+    /// </summary>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1710:IdentifiersShouldHaveCorrectSuffix")]
+    [NonValidatingParameterBinding]
+    public class DeltaSet<TStructuralType> : Collection<IDeltaSetItem>, IDeltaSet where TStructuralType : class
+    {        
+        private Type _clrType;
+        IList<string> _keys;
+               
+        internal IDictionary<string, IDictionary<string, object>> dictKeyValues;
+        internal string currentPathValue;
+
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DeltaSet{TStructuralType}"/> class.
+        /// </summary>
+        public DeltaSet()            
```

Comment:
>DeltaSet [](start = 15, length = 8)

You have two constructors (one that doesn't take keys and one that takes keys but doesn't validate it's non-null), but later (i.e., line 111) you assume that keys is non-null.  

When you create the DeltaSet you have an entity type, which you get the keys from and then pass to the second constructor.

Can we remove this default constructor, replace the constructor on 43 with one that takes the entity type, and put the logic in that constructor for determining the keys from the entity type (and returning an error if the keys can't be determined)?

Reply:
Removed the default constructor, but we cant use EdmEntityType instead, remember we had that and removed due to the reason that we shouldnot have anything related to EDM for Typed ones (since its getting type)